### PR TITLE
feat: additional HTTP validation steps

### PR DESCRIPTION
**Pull request by cbrea**
Implemented a couple of HTTP validation steps:

- I store the field value of the json HTTP response "yyyy" in the context storage key "xxxx"
- the HTTP request body based in the JSON

### CODE REVIEW

File: steps/http/steps.go [hunk 1]
```diff
@@ -80,5 +84,8 @@ func (s Steps) InitializeSteps(ctx context.Context, scenCtx *godog.ScenarioConte
 	scenCtx.Step(`^the HTTP response body must be empty$`, func() error {
 		return session.ValidateResponseBodyEmpty(ctx)
 	})
+	scenCtx.Step(`^I store the field value of the json HTTP response "([^"]*)" in the context storage key "([^"]*)"$`, func(key string, ctxtKey string) error {
```
Comment (jlorgal) [inline steps/http/steps.go:87]: To be consistent with common steps, I would rename the step text to:
`I store the element "([^"]*)" from the JSON HTTP response body in context "([^"]*)"$`

Note also that we are writing JSON instead of json. It's better to maintain the coherence.

File: steps/http/steps.go [hunk 2]
```diff
@@ -61,6 +61,10 @@ func (s Steps) InitializeSteps(ctx context.Context, scenCtx *godog.ScenarioConte
 		}
 		return session.ConfigureRequestBodyJSONProperties(ctx, props)
 	})
+	scenCtx.Step(`^the HTTP request body based in the JSON$`, func(message *godog.DocString) error {
```
Comment (jlorgal) [inline steps/http/steps.go:64]: I'd rename it to `^the HTTP request body with the JSON$`

File: test/acceptance/features/http.feature [hunk 3]
```diff
@@ -43,3 +43,31 @@ Feature: HTTP client
...
+          { "attribute": "attribute1", "value": "value1"},
+          { "attribute": "attribute2", "value": "value2"}
+        ]
+      }
+      """
+     When I send a HTTP "POST" request
+     Then the HTTP status code must be "200"
+      And the HTTP response body must comply with the JSON schema "test-schema"
+      And I store the field value of the json HTTP response "json.list.2.attribute" in the context storage key "key"
+      And I store the field value of the json HTTP response "json.list.2.value" in the context storage key "value"
```
Comment (mariolg) [inline test/acceptance/features/http.feature:66]: Better to move them after they have been checked to exist.

File: steps/http/session.go [hunk 4]
```diff
@@ -110,6 +110,12 @@ func (s *Session) ConfigureRequestBodyJSONProperties(ctx context.Context, props
 	return nil
 }
 
+// ConfigureRequestBodyJSONText sets a json string as the HTTP request's body .
```
Comment (jlorgal) [inline steps/http/session.go:113]: Rephrase it to `ConfigureRequestBodyJSONText writes the body in the HTTP request as a JSON from text.`

File: test/acceptance/features/http.feature [hunk 5]
```diff
@@ -43,3 +43,31 @@ Feature: HTTP client
...
+      And the HTTP response body must have the JSON properties
+          | json.list.0.attribute | attribute0   |
+          | json.list.0.value     | value0       |
+          | json.list.1.attribute | attribute1   |
+          | json.list.1.value     | value1       |
+          | json.list.2.attribute | [CTXT:key]   |
+          | json.list.2.value     | [CTXT:value] |
```
Comment (mariolg) [inline test/acceptance/features/http.feature:73]: I think it's a bit weird to test it this way. I propose to:

1. Validate the reponse body normally:
`      And the HTTP response body must have the JSON properties
          | json.list.0.attribute | attribute0   |
          | json.list.0.value     | value0       |
          | json.list.1.attribute | attribute1   |
          | json.list.1.value     | value1       |
          | json.list.2.attribute | attribute2   |
          | json.list.2.value     | value2 |`
2. Then I store in context some keys, as in your example.
3. Send another HTTP request with this values from context.
4. Then  check again the response.

**Review by mariolg — CHANGES_REQUESTED** [review]
Have a look at my comment.

File: steps/http/session.go [hunk 6]
```diff
@@ -187,3 +193,11 @@ func (s *Session) ValidateResponseBodyEmpty(ctx context.Context) error {
 	}
 	return fmt.Errorf("The response body is not empty")
 }
+
+// StoreResponseBodyJSONPropertyValue stores in the context a key value received in the HTTP Response's body.
```
Comment (jlorgal) [inline steps/http/session.go:197]: I'd rename the method to `StoreResponseBodyJSONPropertyInContext`.
I'd also rephrase the comment to: `StoreResponseBodyJSONPropertyInContext extracts a JSON property from the HTTP response body and stores it in the context.`

File: steps/http/session.go [hunk 7]
```diff
@@ -110,6 +110,12 @@ func (s *Session) ConfigureRequestBodyJSONProperties(ctx context.Context, props
 	return nil
 }
 
+// ConfigureRequestBodyJSONText sets a json string as the HTTP request's body .
+func (s *Session) ConfigureRequestBodyJSONText(ctx context.Context, message string) error {
+	s.Request.RequestBody = []byte(message)
```
Comment (jlorgal) [inline steps/http/session.go:115]: I guess that we should force the `Content-Type` HTTP header to `application/json` in both ConfigureRequestBodyJSONText and ConfigureRequestBodyJSONProperties.
Unless we want to add explicitly this HTTP header.
So, my proposal would be:
```
func (s *Session) ConfigureRequestBodyJSONText(ctx context.Context, message string) error {
	s.Request.RequestBody = []byte(message)
	if s.Request.Headers == nil {
		s.Request.Headers = make(map[string][]string)
	}
	s.Request.Headers["Content-Type"] = "application/json"
	return nil
}
```

and you can reuse this functionality from ConfigureRequestBodyJSONProperties to avoid repetition:

```
func (s *Session) ConfigureRequestBodyJSONProperties(ctx context.Context, props map[string]interface{}) error {
	var json string
	var err error
	for key, value := range props {
		if json, err = sjson.Set(json, key, value); err != nil {
			return fmt.Errorf("Error setting property '%s' with value '%s' in the request body. %s", key, value, err)
		}
	}
	return s. ConfigureRequestBodyJSONText(ctx, json)
}
```

File: test/acceptance/features/http.feature [hunk 8]
```diff
@@ -43,3 +43,31 @@ Feature: HTTP client
           | json.inactive         | [FALSE]                   |
           | json.empty            | [EMPTY]                   |
           | json.null             | [NULL]                    |
+  
+  Scenario: Send a POST request defined by a json string
+    Given the HTTP endpoint "[CONF:url]/anything"
+      And the HTTP path "/test3"
+      And the HTTP request headers
+          | Authorization | Bearer access-token |
```
Comment (jlorgal) [inline test/acceptance/features/http.feature:51]: You haven't written the Content-Type HTTP header. See my previous comment to make it implicit with the JSON body step.

**Review by jlorgal — CHANGES_REQUESTED** [review]

File: steps/http/session.go (hunk 7 above)
Comment (cbrea) [reply steps/http/session.go:114]: Good approach, implemented based in your example with a single change: include  s.Request.Headers["Content-Type"] = "application/json"  within the IF block

File: steps/http/session.go (hunk 6 above)
Comment (cbrea) [reply steps/http/session.go]: Done

File: steps/http/session.go (hunk 4 above)
Comment (cbrea) [reply steps/http/session.go]: Done

File: steps/http/steps.go (hunk 2 above)
Comment (cbrea) [reply steps/http/steps.go]: Done

File: steps/http/steps.go (hunk 1 above)
Comment (cbrea) [reply steps/http/steps.go]: OK. Done

File: test/acceptance/features/http.feature (hunk 5 above)
Comment (cbrea) [reply test/acceptance/features/http.feature]: OK.  Created a long scenario but agree that its more clear

File: test/acceptance/features/http.feature (hunk 8 above)
Comment (cbrea) [reply test/acceptance/features/http.feature]: Done, added the implicit indication to clarify

File: steps/http/session.go (hunk 7 above)
Comment (jlorgal) [reply steps/http/session.go:114]: Your proposal is wrong.
If you have a step with headers, then you will not include the content-type header.
Apart from that, append doesn't make sense (and is more verbose) because there is no headers. It wouldn't either make sense to add multiple Content-Type headers.
I notice a typo in my proposal:
```
	if s.Request.Headers == nil {
		s.Request.Headers = make(map[string][]string)
	}
	s.Request.Headers["Content-Type"] = []string{"application/json"}
```

File: test/acceptance/features/http.feature [hunk 9]
```diff
@@ -43,3 +43,44 @@ Feature: HTTP client
...
+      And the HTTP response body must have the JSON properties
+          | json.list.0.attribute | attribute0 |
+          | json.list.0.value     | value0     |
+          | json.list.1.attribute | attribute1 |
+          | json.list.1.value     | value1     |
+          | json.list.2.attribute | attribute2 |
+          | json.list.2.value     | value2     |
+      And I store the element "json.list.0.attribute" from the JSON HTTP response body in context "attr"
+      And I store the element "json.list.0.value" from the JSON HTTP response body in context "val"
+    Then  the HTTP endpoint "[CONF:url]/anything"
```
Comment (jlorgal) [inline test/acceptance/features/http.feature:74]: Fix alignment

File: steps/http/session.go (hunk 7 above)
Comment (cbrea) [reply steps/http/session.go:114]: Understood the logic now:  Not a global constraint but related only to JSON steps
Done

File: test/acceptance/features/http.feature (hunk 9 above)
Comment (cbrea) [reply test/acceptance/features/http.feature]: Done

**Review by jlorgal — APPROVED** [review]
LGTM
